src/utils/lock.js

```diff
@@ -0,0 +1,40 @@
+const EventEmitter = require('events')
+const { KafkaJSError } = require('../errors')
+
+module.exports = class Lock {
+  constructor({ timeout = 1000 } = {}) {
+    this.locked = false
+    this.timeout = timeout
+    this.emitter = new EventEmitter()
+  }
+
+  async acquire() {
+    return new Promise((resolve, reject) => {
+      if (!this.locked) {
+        this.locked = true
+        return resolve()
+      }
+
+      let timeoutId
+      const tryToAcquire = () => {
+        if (!this.locked) {
+          this.locked = true
+          clearTimeout(timeoutId)
+          this.emitter.removeListener('releaseLock', tryToAcquire)
+          return resolve()
+        }
+      }
+
+      this.emitter.on('releaseLock', tryToAcquire)
+      timeoutId = setTimeout(
+        () => reject(new KafkaJSError('Timeout while acquiring lock')),
+        this.timeout
```

Comment:
Maybe this should be exposed as a configuration, since the expected timeout could be different between clusters.

Reply:
a3a9893